rfd/0100-proxy-ssh-grpc.md

```diff
@@ -0,0 +1,197 @@
+by the Proxy could perform the same operations as the existing SSH server
+but without as much overhead required to establish the session. To minimize
+changes both in `tsh` and on Cluster admins, the existing SSH port can be multiplexed
+to accept both SSH and gRPC by leveraging the TLS ALPN protocol `teleport-proxy-grpc-ssh`.
+Any incoming requests on the SSH listener with said ALPN protocol will be routed
+to the gRPC server and all other requests to the SSH server.
+
+Note: a gRPC server is already exposed via the Proxy web address that users the ALPN protocol
+`teleport-proxy-grpc`. In order to not conflict the new ALPN protocol is required. Reusing the
+existing gRPC server is not an option since it is unauthenticated and has aggressive keep alive
+parameters.
+
+### Proto Definition
+
+The specification is modeled after the [ProxyService](https://github.com/gravitational/teleport/blob/master/api/proto/teleport/legacy/client/proto/proxyservice.proto)
+which is a similar transport mechanism leveraged for Proxy Peering.
+
+```proto
+service ConnectionProxyService {
+  // GetClusterDetails provides cluster information from the perspective of the Proxy
+  rpc GetClusterDetails(GetClusterDetailsRequest) returns (GetClusterDetailsResponse);
+  // DialNode establishes a connection to the target host
+  // 
+  // The bidirectional stream is a transport mechanism to send raw data between
+  // client and server. This allows other protocols(SSH) to be sent through the
+  // tunnel.
+  rpc DialNode(stream DialNodeRequest) returns (stream DialNodeResponse);
+}
+
+// Request for DialNode
+//
+// Connection flow
+// -> Target (client)
+// <- Connection (server)
+// <-> Frame/AgentFrame (client/server)
+// 
+// Separating the Frame and AgentFrame allows
+// SSH and the SSH Agent Protocol to be multiplexed
+// over the same connection.
+message DialNodeRequest {
+  oneof payload {
+    // Contains the information to identify and connect
+    // to the target host
+    Target dial_target = 1;
+    // Raw data to transmit
+    DataFrame frame = 2;
+  }
+}
+
+// Response for DialNodeHost
+message DialNodeResponse {
+  oneof payload {
+    // Cluster configuration details
+    ClusterDetails details = 1;
+    // Raw data to transmit
+    DataFrame frame = 2;
+  }
+}
+
+// Protocol represents the supported protocols
+// that can be proxied.
+enum Protocol {
+  PROTOCOL_UNSPECIFIED = 0;
+  // SSH
+  PROTOCOL_SSH = 1;
+  // SSH Agent
+  PROTOCOL_SSH_AGENT = 2;
+}
+
+// A data frame carrying the payload and 
+// procotol. Allows for multiplexing different
+// protocols on the same stream.
+message DataFrame {
+  // The protocol that sent the data
+  Protocol protocol = 1;
+  // The raw packet of data 
+  bytes payload = 2;
+}
+
+// Target indicates which server to connect to
+message Target {
+  // The server hostname/ip/uuid
+  string host = 1;
+  // The cluster the server is a member of
+  string cluster = 2;
+  // The OS login 
+  string login = 3;
```

Comment:
What's this for?

Reply:
In proxy recording mode, the OS login will be sent as part of the SSH handshake with the proxy, and outside of proxy recording mode the actual OS login that's going to be used in SSH is not exposed to the proxy, so the client could just lie about this.

Reply:
IOW, when not in proxy recording mode, the client could literally just hardcode `-teleport-internal-join` here and the proxy would have to let it through.